mastermind.py

```diff
@@ -0,0 +1,125 @@
+import random
+import string
+import matplotlib.pyplot as plt
+
+def generateRandomString(length):
+    characters = string.ascii_letters + string.digits + string.punctuation
+    randomString = ''.join(random.choice(characters) for _ in range(length))
+    return randomString
+
+def fitnessFunction(candidate, target):
+    return sum(1 for a, b in zip(candidate, target) if a == b)
+
+def selectParents(population, fitnessScores, numParents):
+    parents = sorted(zip(population, fitnessScores), key=lambda x: x[1], reverse=True)[:numParents]
+    return [p[0] for p in parents]
+
+def crossover(parent1, parent2):
+    crossoverPoint = random.randint(0, len(parent1) - 1)
+    return parent1[:crossoverPoint] + parent2[crossoverPoint:]
+
+def mutate(candidate, mutationRate):
+    characters = string.ascii_letters + string.digits + string.punctuation
+    candidateList = list(candidate)
+    for i in range(len(candidateList)):
+        if random.random() < mutationRate:
+            candidateList[i] = random.choice(characters)
+    return ''.join(candidateList)
+
+def geneticAlgorithm(testString, populationSize, iterations, elitism, crossoverNum, generation, mutationRate):
+    population = [generateRandomString(len(testString)) for _ in range(populationSize)]
+    
+    for i in range(iterations):
+        fitnessScores = [fitnessFunction(individual, testString) for individual in population]
+        nextPopulation = selectParents(population, fitnessScores, elitism)
+
+        nonGeneratedPopulationSize = populationSize - generation
+        
+        while len(nextPopulation) < nonGeneratedPopulationSize:
+            parents = random.sample(nextPopulation[:elitism], 2)
+            for _ in range(crossoverNum):
+                offspring = crossover(parents[0], parents[1])
+                offspring = mutate(offspring, mutationRate)
+                nextPopulation.append(offspring)
+        
+        nextPopulation = nextPopulation[:nonGeneratedPopulationSize] + [generateRandomString(len(testString)) for _ in range(generation)]
+        population = nextPopulation
```

Comment:
kicsit korulmenyes logika. Ha jol ertem, `generation` darab randomot akarsz beletenni minden uj generacioba (akkor lehetne akar erre utalo neve is). Ha igy van, egyszerubb lenne elobb belerakni oket az elitizmus utan, aztan utana menni addig, amig a popsize megvan-e.

Reply:
Fair point, abban gondolkoztam, hogy a legjobbak vannak elöl, utána a kicsit gyengébb gyerekeik, aztán a maradék random, de valóban a generálási logika így eléggé meg lett csavarva.